### PR TITLE
Performance improvements for pull request list API

### LINKED ISSUE #30483: Performance issue on pulls list API endpoint with lots of PR's
### Description

We're running renovatebot and were getting timeouts (and renovatebot crashes) because the api/v1/repos/{owner}/{repo}/pulls?state=all endpoint is taking too long on repos with lots of PR's > 3k ( this takes almost 2 minutes, also the repo's are on NFS)

I've traced the issue to the `ToAPIPullRequest` function https://github.com/go-gitea/gitea/blob/b84baf21fa19521e1ab303a60918c74f85fcad1c/services/convert/pull.go#L23

That function is doing `git.OpenRepository` a couple of times, which for > 3k pr's means it doing this ~ 10k times.
I'm not sure how to solve this, maybe it's possible to directly add extra keys to the `PullRequest` model so they don't need to be queried live or adding a separate cache for this ?

### Gitea Version

1.21.10

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Using gitea docker image

### Database

MySQL/MariaDB

### PR DESCRIPTION
**Pull request by lunny**
Fix #30483

### CODE REVIEW

File: routers/api/v1/repo/pull.go [hunk 1]
```diff
@@ -116,23 +116,39 @@ func ListPullRequests(ctx *context.APIContext) {
...
 
 	apiPrs := make([]*api.PullRequest, len(prs))
+	// NOTE: load repository first, so that issue.Repo will be filled with pr.BaseRepo
+	if err := prs.LoadRepositories(ctx); err != nil {
+		ctx.Error(http.StatusInternalServerError, "LoadAttributes", err)
+		return
+	}
+	issueList, err := prs.LoadIssues(ctx)
+	if err != nil {
+		ctx.Error(http.StatusInternalServerError, "LoadAttributes", err)
```
Comment (yp05327) [inline routers/api/v1/repo/pull.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "LoadIssues", err)
```

**Comment by yp05327** [comment]
Does web side have same issue?

**Comment by lunny** [comment]
> Does web side have same issue?

I don't think web router of issue/pulls has the same issue.

**Review by silverwind — APPROVED** [review]

File: models/issues/assignees.go [hunk 2]
```diff
@@ -35,6 +39,7 @@ func (issue *Issue) LoadAssignees(ctx context.Context) (err error) {
 		Join("INNER", "issue_assignees", "assignee_id = `user`.id").
 		Where("issue_assignees.issue_id = ?", issue.ID).
 		Find(&issue.Assignees)
+	issue.isAssigneeLoaded = true
```
Comment (yp05327) [inline models/issues/assignees.go]: IMO, if error occurs, we should not set this flag as true.

Comment (lunny) [reply models/issues/assignees.go]: bb45ecb70754c0f5530e809a42b2a678b68299a1

Comment (yp05327) [reply models/issues/assignees.go]: not only here, but also other places
Are there any reason for setting isXXXLoaded = true before handling error? 🤔

Comment (lunny) [reply models/issues/assignees.go]: e1ee0aa

File: models/issues/issue.go [hunk 3]
```diff
@@ -96,43 +96,47 @@ func (err ErrIssueWasClosed) Error() string {
...
+	Labels            []*Label               `xorm:"-"`
+	isLabelsLoaded    bool                   `xorm:"-"`
+	MilestoneID       int64                  `xorm:"INDEX"`
+	Milestone         *Milestone             `xorm:"-"`
+	isMilestoneLoaded bool                   `xorm:"-"`
+	Project           *project_model.Project `xorm:"-"`
+	Priority          int
+	AssigneeID        int64            `xorm:"-"`
+	Assignee          *user_model.User `xorm:"-"`
+	isAssigneeLoaded  bool             `xorm:"-"`
```
Comment (yp05327) [inline models/issues/issue.go:122]: A question: `Assignees` is same to `Assignee`, so we only need `isAssigneeLoaded` and no `isAssigneesLoaded`?

File: models/issues/issue_list.go [hunk 4]
```diff
@@ -310,6 +312,7 @@ func (issues IssueList) loadAssignees(ctx context.Context) error {
 
 	for _, issue := range issues {
 		issue.Assignees = assignees[issue.ID]
+		issue.isAssigneeLoaded = true
```
Comment (yp05327) [inline models/issues/issue_list.go:316]: If I understand correctly, in issue.LoadAssignees we will set issue.assignee, but in issuelist.LoadAssignees we will not do this. is it correct?
https://github.com/go-gitea/gitea/pull/30490/files#diff-d3a8b9a325e353adebf461f60e7a34be0be36306a8f83f56d3c8cf372f24c346R46-R49

File: models/issues/pull.go [hunk 5]
```diff
@@ -289,18 +290,18 @@ func (pr *PullRequest) LoadHeadRepo(ctx context.Context) (err error) {
...
 
 	reviews, err := GetReviewsByIssueID(ctx, pr.Issue.ID)
 	if err != nil {
 		return err
 	}
-
 	if err = reviews.LoadReviewers(ctx); err != nil {
 		return err
 	}
+	pr.isRequestedReviewersLoaded = true
```
Comment (yp05327) [inline models/issues/pull.go:304]: maybe move to L308 is better.

File: models/issues/pull_list.go [hunk 6]
```diff
@@ -153,50 +155,93 @@ func PullRequests(ctx context.Context, baseRepoID int64, opts *PullRequestsOptio
...
+	return repoIDs.Values()
+}
+
+func (prs PullRequestList) LoadRepositories(ctx context.Context) error {
+	repoIDs := prs.getRepositoryIDs()
+	reposMap := make(map[int64]*repo_model.Repository, len(repoIDs))
+	if err := db.GetEngine(ctx).
+		In("id", repoIDs).
+		Find(&reposMap); err != nil {
+		return fmt.Errorf("find issues: %w", err)
```
Comment (yp05327) [inline models/issues/pull_list.go]: find repos

File: routers/api/v1/repo/pull.go [hunk 7]
```diff
@@ -116,23 +116,39 @@ func ListPullRequests(ctx *context.APIContext) {
 	}
 
 	apiPrs := make([]*api.PullRequest, len(prs))
+	// NOTE: load repository first, so that issue.Repo will be filled with pr.BaseRepo
+	if err := prs.LoadRepositories(ctx); err != nil {
+		ctx.Error(http.StatusInternalServerError, "LoadAttributes", err)
```
Comment (yp05327) [inline routers/api/v1/repo/pull.go]: LoadRepositories

File: models/issues/pull_list.go [hunk 8]
```diff
@@ -153,50 +155,93 @@ func PullRequests(ctx context.Context, baseRepoID int64, opts *PullRequestsOptio
...
+		}
+		if pr.HeadRepo == nil {
+			pr.HeadRepo = reposMap[pr.HeadRepoID]
+			pr.isHeadRepoLoaded = true
+		}
+	}
+	return nil
+}
+
 func (prs PullRequestList) LoadAttributes(ctx context.Context) error {
```
Comment (yp05327) [inline models/issues/pull_list.go:191]: LoadAttributes will only load issues? seems a little strange. 🤔

File: models/issues/issue.go (hunk 3 above)
Comment (lunny) [reply models/issues/issue.go:122]: Yes. Both Assignees and Assignee are loaded in the `LoadAssignees`, so we don't need two variables.

File: models/issues/pull_list.go (hunk 8 above)
Comment (yp05327) [reply models/issues/pull_list.go:191]: You added func ResetAttributesLoaded, so at least LoadAttributes should load the attributes in rest function?
Then the usage of LoadAttributes before these changes, should rename to LoadIssues?

File: models/issues/pull.go (hunk 5 above)
Comment (lunny) [reply models/issues/pull.go:304]: They are the same.

File: models/issues/pull_list.go (hunk 8 above)
Comment (lunny) [reply models/issues/pull_list.go:191]: I think so. But I will not touch that to bloat this PR.

File: models/issues/issue_list.go (hunk 4 above)
Comment (lunny) [reply models/issues/issue_list.go:316]: d21752894906e7cdd92753cf9cb5c453412e5ab8

File: models/issues/pull_list.go (hunk 6 above)
Comment (lunny) [reply models/issues/pull_list.go]: d21752894906e7cdd92753cf9cb5c453412e5ab8

File: routers/api/v1/repo/pull.go (hunk 7 above)
Comment (lunny) [reply routers/api/v1/repo/pull.go]: d21752894906e7cdd92753cf9cb5c453412e5ab8

File: services/convert/pull.go [hunk 9]
```diff
@@ -44,7 +45,16 @@ func ToAPIPullRequest(ctx context.Context, pr *issues_model.PullRequest, doer *u
...
+	var doerID int64
+	if doer != nil {
+		doerID = doer.ID
+	}
+
+	const repoDoerPermCacheKey = "repo_doer_perm_cache"
+	p, err := cache.GetWithContextCache(ctx, repoDoerPermCacheKey, fmt.Sprintf("%d_%d", pr.BaseRepoID, doerID),
+		func() (access_model.Permission, error) {
+			return access_model.GetUserRepoPermission(ctx, pr.BaseRepo, doer)
+		})
```
Comment (yp05327) [inline services/convert/pull.go:57]: You will always read cache no matter doer is nil or not.
So if doer is not nil, then read cache, or set the permission to none?

**Comment by yp05327** [comment]
image
LoadLabels can be moved to L64-L66?

File: models/issues/issue_list.go [hunk 10]
```diff
@@ -74,11 +74,11 @@ func (issues IssueList) LoadRepositories(ctx context.Context) (repo_model.Reposi
 
 func (issues IssueList) getPosterIDs() []int64 {
 	return container.FilterSlice(issues, func(issue *Issue) (int64, bool) {
-		return issue.PosterID, true
+		return issue.PosterID, issue.Poster == nil
```
Comment (yp05327) [inline models/issues/issue_list.go]: I don't agree this change, as function name with `getXXXIDs` will always return all xxx's IDs in other places, but not IDs which are not loaded. Others may confused with it.

File: services/convert/pull.go (hunk 9 above)
Comment (lunny) [reply services/convert/pull.go:57]: `doer` has been checked in `GetUserRepoPermission`.

File: models/issues/issue_list.go (hunk 10 above)
Comment (lunny) [reply models/issues/issue_list.go]: 54e8493e451354e908d86617d190db23896efb14

**Review by yp05327 — APPROVED** [review]

**Pull request merged**